parser-typechecker/src/Unison/Hashing/V2/ABT.hs

```diff
@@ -78,27 +79,25 @@ hash :: forall f v a h . (Functor f, Hashable1 f, Eq v, Show v, Ord h, Accumulat
      => Term f v a -> h
 hash = hash' [] where
   hash' :: [Either [v] v] -> Term f v a -> h
-  hash' env (Term _ _ t) = case t of
-    Var v -> maybe die hashInt ind
+  hash' env = \case
+    Var' v -> maybe die hashInt ind
       where lookup (Left cycle) = v `elem` cycle
             lookup (Right v') = v == v'
             ind = findIndex lookup env
             hashInt :: Int -> h
             hashInt i = Hashable.accumulate [Hashable.Nat $ fromIntegral i]
             die = error $ "unknown var in environment: " ++ show v
                         ++ " environment = " ++ show env
-    Cycle (AbsN' vs t) -> hash' (Left vs : env) t
-    -- Cycle t -> hash' env t
-    Abs v t -> hash' (Right v : env) t
-    Tm t -> Hashable.hash1 (hashCycle env) (hash' env) t
+    Cycle' vs t -> Hashable.hash1 (hashCycle vs env) undefined t
+    Abs'' v t -> hash' (Right v : env) t
+    Tm' t -> Hashable.hash1 (\ts -> (List.sort (map (hash' env) ts), hash' env)) (hash' env) t
```

Comment:
Hrmm, should this sort maybe go into the `Hashable1` instances instead? It seems at a glance like that'd be the right place for it, you'd use the provided hashing function to hash the list of components, then you'd sort the returned list of hashes within the Hashable1 instance before using them.

Seems like the ordering is more a property of the container than of the hashing function 🤔 

I don't think I have all the context here, but let me know what you think of that.

Reply:
I think that's the opposite opinion that @pchiusano had - we decided the `Hashable1` instances should not be sorting any hashes themselves, but rather receive a sorted list of hashes from the result of `hashCycle`.

Reply:
@ChrisPenner - right, what @mitchellwrosen said. Normalizing the order of an unordered set of things is something that multiple `Hashable1` instances want to do. As it's a bit finnicky, we want to do it in a standardized way in a single place, and have that logic be wrapped up in the `hashCycle` argument that's passed to the `Hashable1` instance.

Reply:
> Normalizing the order of an unordered set of things is something that multiple Hashable1 instances want to do. As it's a bit finnicky, we want to do it in a standardized way in a single place, and have that logic be wrapped up in the hashCycle argument that's passed to the Hashable1 instance.

Definitely understand the benefits of putting this somewhere where it affects everything so that you can't "forget" to do it in a new instance, but this will also prevent us from having any Hashable1 containers which _are_ order dependent won't it? E.g. as a thought experiment, we add anonymous products down the road and this is a Hashable1 container, but the order of the fields is important and should be part of the hash; wouldn't this change prevent us from using the ordering in that hash?